### ISSUE TITLE
Theory question on warmup

### ISSUE THREAD

**Issue by OverLordGoldDragon** [LiyuanLucasLiu/RAdam#22]
> _Due to the lack of samples in the early stage, the adaptive learning rate has an_
> _undesirably large variance, which leads to suspicious/bad local optima_ -- pg. 3

Does this apply when feeding the same dataset in a _different configuration_? Namely, I'm training a timeseries (16-channel EEG) CNN-LSTM classifier, and vary the input `timesteps` across epochs for the _same model_. While the information source probability distribution remains identical, what the neural net effectively "sees" differs substantially between, say, 13500 and 216000 `timesteps`. 

This considered, is warmup for the first epoch of every new `timesteps` setting advisable? Thanks

**Comment by LiyuanLucasLiu** [comment]
Thanks for asking :- ) But I'm sorry that i'm not sure i understand this question. It will be better if you can elaborate more : -)

**Comment by OverLordGoldDragon** [comment]
@LiyuanLucasLiu Sure: my model architecture is `CNN --> LSTM` - both are able to deal with _variable_ `timesteps`. Instead of training to classify _N_ `timesteps`, I vary `timesteps` as a form of 'warmup' and data augmentation before reaching _N_ - so if _N_=216000, for example, I start with `timesteps=216000//16`, then `//8`, `//4`, `//2`, and finally `216000`. While the data's the same, the information that the model receives _per update/batch_ is different - thus, I'd suspect the optimizer would benefit from warmup on each new configuration to lower variance.

_However_, I experimented with `RAdam` vs `Adam` and `RNAdam` vs `NAdam` - the results are: `R`'s substantially lower validation loss variance and prevent spikes, but non-`R`'s yield better final loss and F1-score. 

What I did not test: `Adam`/`Nadam` _with_ warmup, or `RAdam`/`RNAdam` _without_ warmup. It's possible that lower variance isn't beneficial in the long run, similar to `SGD` - due to converging to local minima prematurely. 

Lastly: it's possible the [implementations](https://github.com/CyberZHG/keras-radam) were off, or that my warmup wasn't configured properly (one full epoch or 1/2 epoch). But as is, `R` was bested by non-`R` consistently across many hyperparameters.

**Comment by LiyuanLucasLiu** [comment]
Thanks for letting us know : -) In this situation, I don't think there is a firm answer to the question. From my point of view, these stages are similar to each other, and the gradient statistics should be relative similar. 

From the practical point of view, I think warmup may not be really necessary, since the vanilla adam works reasonable well -- it means the variance issue is not very serious (as in Figure 9 in the paper, the magnitude of the variance can be negligible in some situation). In this situation, I would recommend to:
1. use RAdam / RNAdam without warmup: using additional warmup may not be necessary, since the vanilla Adam works well;
2. tune the hyper-parameters (especially learning rates / learning rate scheduler) a little bit; 

I myself didn't use tensorflow (before pytorch, i mainly use torch + lua XD), and i'm not 100% sure that it is correct (also, I just fixed a bug in my pytorch implementation, as in: [this commit](https://github.com/LiyuanLucasLiu/RAdam/commit/1d146e572c0e0f170cd7a7bf5995873ccc4768d0)). 

Hope it helps.

**Comment by OverLordGoldDragon** [comment]
@LiyuanLucasLiu Understood - thanks for the response, and for your work
